main_test.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"reflect"
+	"strconv"
+	"testing"
+)
+
+func Test_sortInterfaces(t *testing.T) {
+	testCases := []struct {
+		interfaces       []string
+		sortedInterfaces []string
+	}{
+		{
+			interfaces:       []string{"eth0"},
+			sortedInterfaces: []string{"eth0"},
+		},
+		{
+			interfaces:       []string{"eth1"},
+			sortedInterfaces: []string{"eth1"},
+		},
+		{
+			interfaces:       []string{"eth0", "eth1"},
+			sortedInterfaces: []string{"eth0", "eth1"},
+		},
+		{
+			interfaces:       []string{"eth1", "eth0"},
+			sortedInterfaces: []string{"eth0", "eth1"},
+		},
+		{
+			interfaces:       []string{"bond1", "bond0"},
+			sortedInterfaces: []string{"bond0", "bond1"},
+		},
+		{
+			interfaces:       []string{"eth1", "bond0"},
+			sortedInterfaces: []string{"bond0", "eth1"},
+		},
+		{
+			interfaces:       []string{"ens256", "ens224"},
+			sortedInterfaces: []string{"ens224", "ens256"},
+		},
+		{
+			interfaces:       []string{"eth4", "eth1", "eth0"},
+			sortedInterfaces: []string{"eth0", "eth1", "eth4"},
+		},
```

Comment:
I think this would fail:

```go
		{
			interfaces:       []string{"eth10", "eth2"},
			sortedInterfaces: []string{"eth2", "eth10"},
		},
```

Reply:
that is true this would not work, I will check if this can ever happen

Reply:
So on AWS, there are few instance types (16xlarge,18xlarge,24xlarge) which can have up to 15 Network interfaces but the rest can have only up to 8. So this could potentially happen, However, we only allow 110 pods per machine(each pod  gets 1 IP) and the IP limit per interface on such instances is so big (50 ips per ENI) that we would only need 3 or 4 interfaces to fill the pod limit so the scenario with 11 interfaces  (`eth10`) should not happen in our clusters.
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI

IMHO We can ignore cases where we would have 11 interfaces as that cannot happen now.

Reply:
IMO we should invest the twenty minutes now to make it more resilient instead of waiting until something happens where it breaks and we spend days again to figure out what is going on and how to fix it at a later point in time.

Reply:
I don't have an idea how to quickly solve it now.  So that's why I tried to assume how probable is that we would ever get to that scenario. Currently, it does not seem to be possible.

 I don't see the scenario of having more than 10 interfaces likely ever since that would mean we would run around 450-600 pods per node which is something we never wanna do since other components are starting to break on such scenarios ( like systemd, networking, and others).

So if anyone of you has a  viable solution please share it with me. Otherwise, I would be fine with going with this.

Reply:
and as Pawel mentioned the `eth0` will always be first no matter how many interfaces we have

Reply:
When we always only want `eth0`, why are we doing all the semi broken sorting magic? Fixing that sort should be rather simple in go if this is necessary.

Reply:
See e.g. https://golangcode.com/sorting-an-array-of-numeric-items.

Reply:
> When we always only want eth0

we need `INTERFACE_NAME0` but the interface name might  differ on other providers, specially on-prem so we cannot simply hard-code to `eth0`

> See e.g. https://golangcode.com/sorting-an-array-of-numeric-items.

but that's for numbers only? how should it look for mixed string with number? I assume caaling `strconv.Atoi("eth11")` will cause an error

Reply:
Maybe we can pick `^[a-z]0$`?

Reply:
Like Pawel suggested, you just take the number out of the strings so that you can compare the number. When you do it with a regex you can either use a capturing group or you remove anything that is not the number.

Reply:
I'd skip comparison and take interface that matches `^[a-z]+0$` pattern when we know we want the one ending with `0`.

Reply:
That is not possible as well, as there can be  `bond0.2369` for on-prem or `ens112` (for vmware). But only on AWS where we have `ethX` this sorting is needed as only with AWSC NI there will be multiple default routes. And this tool only looks for the `DEFAULT_IP` on the default routes.

Reply:
I would like to move on with this, I still want to do a bunch of testing to ensure this works on all providers and if we don't do some significant changes to our infra and networking this should work fine and it will fix the PM.

Reply:
When you think it makes sense to write a unit test for this then please do that. Otherwise I trust your judgement and you can go ahead as far as I am concerned.